### PR TITLE
[inductor] Preserve logical indices for fused arg reductions

### LINKED ISSUE #193661: [inductor] Wrong argmax after fused trailing-dim reduction on non-contiguous inputs (incl. non-unit-stride reduction window) — generic reduction fusion ignores input strides, CPU and CUDA
### 🐛 Describe the bug

**Relationship to #193492**: this is the same *mechanism* (the fused kernel assumes a contiguous interpretation of the input and ignores the real strides) but a **different lowering** — the generic reduction fusion (`mean`/`var`/`sum`/`logsumexp`/`amax` + `argmax`), not the `_adaptive_avg_pool2d` path. The contiguous guard proposed in #193492 is added at the `_adaptive_avg_pool2d` lowering and does not apply to these graphs; the reduction path needs its own fix/regression coverage. I asked in the #193492 thread whether this path should be tracked separately ([[comment](https://github.com/pytorch/pytorch/issues/193492#issuecomment-5299760647)](https://github.com/pytorch/pytorch/issues/193492#issuecomment-5299760647)); opening this issue here so it doesn't get lost in the pool discussion. Happy to close this and fold everything back into #193492 if maintainers prefer a single tracking issue.

`torch.compile` with the Inductor backend returns a wrong `argmax` index when a trailing-dim reduction (`mean`, `var`, `sum`, `logsumexp`, also `amax`) is fused with `argmax` and the reduction input is a **non-contiguous view**. Eager returns the correct index; the compiled index points at a value that is **not** the extremum of the reduction output. The unfused reduction matches eager exactly (sanity `maxdiff ≈ 0`), so the mis-indexing is introduced by the fused kernel, whose generated input indexing assumes a contiguous layout and ignores the real strides. The trigger is deterministic (fixed seed, no randomness).

The previously-reported family instances used non-contiguous leading dims with a **stride-1** reduction window (e.g. `transpose`). The new instance below additionally has a **non-unit-stride reduction window** (`[..., ::2]`, stride 2 on the reduced dim) combined with non-contiguous leading dims. Neither axis triggers on its own: the same `[..., ::2]` slice on a contiguous input passes, and a fully contiguous input passes. As a scope hint for the regression test, `expand`/stride-0 views are **not** affected (all pass).

### To reproduce

```python
import torch

def fn(x):
    return torch.argmax(torch.mean(x, dim=-1))

torch.manual_seed(123)
x = torch.rand([2, 4, 4, 8]).transpose(0, 1).contiguous().transpose(0, 1)[..., ::2]
# shape (2, 4, 4, 4), strides (32, 64, 8, 2)
# non-contiguous leading dims AND a stride-2 reduction window

print("eager   =", fn(x).item())                # 12 (correct)
print("compiled=", torch.compile(fn)(x).item()) # 24 (wrong)
```

Truth check: value at the returned index vs. the true maximum of `mean(x, dim=-1)` (eager index points at the maximum; compiled index does not):

```text
eager    index 12 -> value 0.828343 (== max)
compiled index 24 -> value 0.596202 (!= max)
```

Scope / additional observations (all fixed seed 123, caches disabled, unfused reductions matching eager exactly):

- The same input mismatches for `var`, `sum`, `logsumexp` (dim=-1) and for `mean`, `sum`, `var` (dim=[-1, -2]): **7/7 reduction variants**, on CPU and CUDA, torch 2.13.0 and 2.5.1 (CPU indices identical across versions, e.g. mean: eager 12 / compiled 24 / true 12; CUDA also reproduces).
- Controls: the same `[..., ::2]` slice on a contiguous input passes, a fully contiguous input passes, and `expand`/stride-0 views pass — the trigger is the combination of non-contiguous leading dims and a non-unit-stride reduction window.
- The earlier family instances (non-contiguous leading dims, stride-1 reduction window) are also affected; see the #193492 thread for the family matrix.

Kernel evidence (CPU, torch 2.13.0): the aten graph contains only `mean.dim -> argmax` (the transpose/slice views are metadata and never enter the graph), while the fused kernel `cpp_fused_argmax_mean_0` reads the reduction window with a flat, contiguous interpretation:

```cpp
for (int64_t x0 = 0; x0 < 32; x0++) {   // x0 = flattened 2*4*4 leading positions
    auto tmp0 = in_ptr0[8 * x0];        // mean over in_ptr0[8*x0 + {0,2,4,6}]
    auto tmp1 = in_ptr0[2 + 8 * x0];
    auto tmp3 = in_ptr0[4 + 8 * x0];
    auto tmp5 = in_ptr0[6 + 8 * x0];
    ...
    tmp_acc0 = argmax_combine(tmp_acc0, mean, x0);
}
```

The true offset of element (i, j, k, k′) with strides `(32, 64, 8, 2)` is `32i + 64j + 8k + 2k′`, which the fused flat reads do not match. The wrong reads stay in bounds (max flat offset 254 < storage size 256), so the result is silently wrong — no OOB, no exception.

### Error logs

No exception — silent wrong result (wrong index returned).

### Versions

```text
PyTorch version: 2.13.0+cu130 and 2.5.1+cu121 (both reproduce, CPU and CUDA)
OS: Linux (AutoDL container)
```

cc @ezyang @gchanan @kadeng @msaroufim @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by SachaKefif**
# Fixing an Issue

Before submitting, please review:
- [The Ultimate Guide to PyTorch Contributions](https://github.com/pytorch/pytorch/wiki/The-Ultimate-Guide-to-PyTorch-Contributions#getting-started-with-pull-requests)
- [AI-Assisted Development](https://github.com/pytorch/pytorch/blob/main/AI_POLICY.md) policy

---

## Issue

Fixes #193661

## Summary

The affected function first computes `mean(x, dim=-1)` and then applies
`argmax` to the result. Inductor reorders iteration over the remaining
dimensions to improve memory access for the non-contiguous input. Although it
examines the correct set of reduced values, the generated reduction reports the
physical reduction-loop index instead of the corresponding logical row-major
index. In the reproducer, the logical index `12` is therefore reported as `24`.

Inductor already computes a separate logical row-major index for
multidimensional `argmax` and `argmin` reductions over permuted or
non-contiguous realized inputs. Here, however, `mean` is fused into `argmax` and
the reduction input is represented by an unrealized `Pointwise` node, which
does not have a layout that the existing check can inspect. Consequently, the
logical-index path was not enabled.

This patch enables the existing logical-index path when the reduction input is
a `StorageBox` containing a `Pointwise` node. It does not change how candidate
values are computed; it supplies their logical indices to the existing
arg-reduction implementation. This may add index arithmetic to fused
multidimensional `argmax` and `argmin` operations, and the performance impact
has not been measured.

The regression test recreates the non-contiguous input from the issue. Because
it is added to `CommonTemplate`, it is instantiated in the device-generic CPU
and GPU test suites and checks compiled output against eager output.

> AI assistance disclosure: I used OpenAI Codex to help inspect the generated IR

## Test plan

- `git diff --check`
- `python -m py_compile torch/_inductor/lowering.py test/inductor/test_torchinductor.py`
- Full compiled regression test not run locally because `cl.exe` is unavailable.

## Checklist

- [ ] Passes lint (`spin fixlint`)
- [x] Added/updated tests
- [ ] Updated documentation (not applicable)
- [ ] Included benchmark results (not applicable; correctness fix)

## BC-breaking?

No

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by jansel** [comment]
@claude review these changes

**Comment by eellison** [comment]
duping my comment from https://github.com/pytorch/pytorch/issues/193661.

what do you think ? cc @jananisriram who is assigned,  @karthickai  who did first version of this pr , to take over

**Comment by sylvesterkaczmarek** [comment]
@eellison @SachaKefif I split out the Triton reduction-cache blocker into #194093. It makes the reduction CSE key include the logical index, matching the C++ codegen, and adds a regression with x.argmax() and x.t().argmax() in the same compiled graph. This should leave #193680 focused on the logical-index propagation/layout fix.

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -18979,6 +18979,33 @@ def f(x):
         self.assertTrue("ReductionHint.OUTER" in code)
         self.assertFalse("ReductionHint.INNER" in code)
 
+    @parametrize("slice_pointwise", (False, True))
+    @skip_if_halide  # Halide reports the physical index after loop reordering
+    def test_argmin_argmax_fused_reduction_logical_index(self, slice_pointwise):
```
Comment (jananisriram) [inline test/inductor/test_torchinductor.py]: Since #193492 was closed as a duplicate on the understanding that this fix covers it, would it be worth pinning that case here too? I checked the branch and both the transposed `adaptive_avg_pool1d + argmax` case and the `channels_last` case from #193751 come out correct, but nothing in the suite would catch it if the guard is ever narrowed again for perf.

Smaller thing: the adjacent `test_triton_argmin_argmax_transpose_logical_index` uses `self.common(fn, (x,))`, which would pick up the cpp_wrapper and dynamic-shapes variants too. The explicit `(5, 12)` assertion could stay on top of that.

[from my agent]

File: benchmarks/dynamo/microbenchmarks/bench_argreduce_logical_index.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import argparse
```
Comment (jananisriram) [inline benchmarks/dynamo/microbenchmarks/bench_argreduce_logical_index.py]: The commit message mentions CUDA benchmarking was not available locally, so this has not produced numbers yet, and nothing in the repo invokes it. Would it be simpler to drop the file and put the numbers in the description instead? I ran the equivalent on an H100 and the contiguous shapes came out flat - happy to share the script and the full table.

[from my agent]

File: test/inductor/test_torchinductor.py (hunk 1 above)
Comment (jananisriram) [inline test/inductor/test_torchinductor.py:18984]: nit: maybe worth using `self.common(fn, (x,))` from `test_triton_argmin_argmax_transpose_logical_index`, to pick up `cpp_wrapper` and dynamic shape variants?

File: benchmarks/dynamo/microbenchmarks/bench_argreduce_logical_index.py (hunk 2 above)
Comment (jananisriram) [inline benchmarks/dynamo/microbenchmarks/bench_argreduce_logical_index.py]: This hasn't produced numbers yet (the commit message notes CUDA wasn't available locally) and nothing in the repo invokes it. Simpler to drop the file and put the numbers in the description? I have H100 numbers if useful.

[from my agent]

File: test/inductor/test_torchinductor.py (hunk 1 above)
Comment (SachaKefif) [reply test/inductor/test_torchinductor.py:18984]: Good point, thanks. I updated the regression test to use self.common(fn, (x,).

**Comment by eellison** [comment]
Could we do the full fix ? https://github.com/pytorch/pytorch/issues/193661#issuecomment-5359104912 

It would be great to do the correctness fix and preserve perf.

**Review by jansel — CHANGES_REQUESTED** [review]
Should we do #195178 instead?
